### PR TITLE
feat(treesitter)!: incremental injection parsing

### PR DESCRIPTION
**Pull request by lewis6991**
### Problem

Treesitter highlighting is slow for large files with lots of injections.

### Solution

Only parse injections we are going to render during a redraw cycle.

---

- `LanguageTree:parse()` will no longer parse injections by default and now requires an explicit `range` argument to be passed.
- `TSHighlighter` now parses injections incrementally during `on_win` callbacks for the redrawn line range.
- Plugins which require certain injections to be parsed must run `parser:parse({ start_row, end_row })` before using the tree.

### CODE REVIEW

File: runtime/doc/treesitter.txt [hunk 1]
```diff
@@ -1167,10 +1167,17 @@ LanguageTree:named_node_for_range({range}, {opts})
...
-    determine if any child languages should be created.
+    tree represents.
+
+    If {range} is provided and isn't `false`, will query and parse injections.
+
+    Parameters: ~
+      • {opts}  TSParseOpts|nil Optional arguments:
+                • range boolean|Range4: Only parse this range in the buffer.
+                  This only affects trees which have an explicit range defined
+                  (injections) and does not affect the root tree.
```
Comment (clason) [inline runtime/doc/treesitter.txt:1180]: This is slightly ambiguous: Does it parse the range, or just _injections_ in the range?

(And the second sentence probably fits better in the preceding paragraph.)

**Comment by clason** [comment]
We probably want to specify the range here, too:
https://github.com/neovim/neovim/blob/3d948a4dc4b2cd3c8d3ac497caf3dfe25adfb90d/runtime/lua/vim/treesitter/_fold.lua#L165
(since `srow`, `erow` are available here).

File: runtime/doc/treesitter.txt (hunk 1 above)
Comment (lewis6991) [reply runtime/doc/treesitter.txt:1180]: The languagetree has "regions". A region can be empty (undefined) which means parse the whole ~buffer~ source. `range` does not affect these undefined regions.

So if you create a language tree and run `set_included_regions`, you've now defined regions to be parsed and thus `range` will be applied.

Injections are always defined regions, so `range` always applies.

I'll try and tidy this up. It was rushed.

**Comment by clason** [comment]
Note to self: this requires changes in `nvim-treesitter` (`main`; I'm considering keeping `master` locked at 0.9.x).

**Comment by clason** [comment]
Only other `parser:parse` I could find outside our tests is 
https://github.com/neovim/neovim/blob/3d948a4dc4b2cd3c8d3ac497caf3dfe25adfb90d/runtime/lua/vim/treesitter/_query_linter.lua#L274
which should be fine (since we don't have any injections for query language).

**Comment by lucario387** [comment]
> Note to self: this requires changes in `nvim-treesitter` (`main`; I'm considering keeping `master` locked at 0.9.x).

If 0.9.x will be the final version for master branch of `nvim-treesitter`, is it okay to just "switch" entirely to new injection format on that branch then?

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 2]
```diff
@@ -348,19 +385,61 @@ function LanguageTree:parse()
...
+
+--- Run a parse on the language tree.
+---
+--- If {range} isn't `false|nil`, then injections will be
+--- queried and parsed recursively.
+---
+---@param range boolean|Range|nil: Parse this range in the parsers source.
+---    This only affects regions in the source which have an explicit
+---    range defined which includes: regions set with `set_included_regions()`,
+---    and, injections.
```
Comment (clason) [inline runtime/lua/vim/treesitter/languagetree.lua:399]:
```suggestion
---@param range boolean|Range|nil: Parse this range in the parser's source.
---    This only affects regions in the source which have an explicit
---    range defined which includes: regions set with `set_included_regions()`
---    and injections.
```

This is a bit hard to parse for me (pun intended). Can you explain what exactly this `{range}` parameter controls, from a practical point of view?

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 3]
```diff
@@ -348,19 +385,58 @@ function LanguageTree:parse()
...
+
+--- Run a parse on the language tree.
+---
+---@param range boolean|Range|nil: Parse this range in the parsers source.
+---    This only affects regions in the source which have an explicit
+---    range defined which includes: regions set with `set_included_regions()`,
+---    and, injections.
+---    Set to `true` to run a complete parse of the source. (This can be slow.)
+---    Set to `false|nil` to only parse regions with undefined ranges (typically
+---    this is the root tree).
```
Comment (clason) [inline runtime/lua/vim/treesitter/languagetree.lua:399]:
```suggestion
--- Recursively parse all regions in the language tree using a |treesitter-parser|
--- for the corresponding language and run injection queries on the parsed trees
--- to determine whether child trees should be created and parsed.
---
--- Any region with empty range (`{}`, typically only the root tree) is always parsed; 
--- otherwise (typically injections) only if it intersects {range} (or if {range} is `true`).
---
---@param range boolean|Range|nil: Parse this range in the parser's source.
---    Set to `true` to run a complete parse of the source (Note: Can be slow!)
---    Set to `false|nil` to only parse regions with empty ranges (typically
---    only the root tree without injections).
```

Omitting mentioning `set_included_regions()` as this is `@private` (for now).

File: runtime/doc/news.txt [hunk 4]
```diff
@@ -61,6 +61,12 @@ The following changes may require adaptations in user config or plugins.
     spaces (but paths themselves may contain spaces now).
   • |'directory'| will no longer remove a `>` at the start of the option.
 
+• |LanguageTree:parse()| will no longer parse injections by default and
+  now requires an explicit range argument to be passed.
+
+  Plugins which require certain injections to be parsed must run
+  `parser:parse({ start_row, end_row })` before using the tree.
```
Comment (clason) [inline runtime/doc/news.txt:68]:
```suggestion
• |LanguageTree:parse()| will no longer parse injections by default and
  now requires an explicit range argument to be passed. If injections are
  required, provide an explicit range via `parser:parse({ start_row, end_row })`.
```

File: runtime/doc/news.txt [hunk 5]
```diff
@@ -69,6 +75,9 @@ The following new APIs and features were added.
 • Performance:
   • 'diffopt' "linematch" scoring algorithm now favours larger and less groups
     https://github.com/neovim/neovim/pull/23611
+  • The treesitter highlighting now parses injections incrementally during
+    screen redraws for the line range being rendered. This should provide
+    significantly improved in large files with lots of injections.
```
Comment (clason) [inline runtime/doc/news.txt:80]:
```suggestion
  • Treesitter highlighting now parses injections incrementally during
    screen redraws only for the line range being rendered. This significantly
    improves performance in large files with many injections.
```

**Review by clason — APPROVED** [review]

File: runtime/lua/vim/treesitter/languagetree.lua [hunk 6]
```diff
@@ -348,19 +389,60 @@ function LanguageTree:parse()
     end
   end
 
+  return query_time
+end
+
+--- Recursively parse all regions in the language tree using a |treesitter-parser|
+--- for the corresponding language and run injection queries on the parsed trees
```
Comment (clason) [inline runtime/lua/vim/treesitter/languagetree.lua:396]:
```suggestion
--- Recursively parse all regions in the language tree using |treesitter-parsers|
--- for the corresponding languages and run injection queries on the parsed trees
```

my bad... 😞

**Comment by zeertzjq** [comment]
There is a memory leak in this PR:
```
= =================================================================
= ==29097==ERROR: LeakSanitizer: detected memory leaks
= 
= Direct leak of 6 byte(s) in 2 object(s) allocated from:
=     #0 0x56310e2d7d9e in __interceptor_malloc (/home/runner/work/neovim/neovim/build/bin/nvim+0xcc2d9e) (BuildId: 9a852b026c43f6da24429b6c21593877c7f7c615)
=     #1 0x56310f0d6746 in try_malloc /home/runner/work/neovim/neovim/src/nvim/memory.c:88:15
=     #2 0x56310f0d692c in xmalloc /home/runner/work/neovim/neovim/src/nvim/memory.c:122:15
=     #3 0x56310ef0cc30 in nlua_ref /home/runner/work/neovim/neovim/src/nvim/lua/executor.c:1298:51
=     #4 0x56310ef0cca0 in nlua_ref_global /home/runner/work/neovim/neovim/src/nvim/lua/executor.c:1307:10
=     #5 0x56310ef29ae0 in parser_set_logger /home/runner/work/neovim/neovim/src/nvim/lua/treesitter.c:733:11
=     #6 0x56310fd7adf5 in lj_BC_FUNCC (/home/runner/work/neovim/neovim/build/bin/nvim+0x2765df5) (BuildId: 9a852b026c43f6da24429b6c21593877c7f7c615)
= 
= SUMMARY: AddressSanitizer: 6 byte(s) leaked in 2 allocation(s).
====================================================================================================
nan ms test/helpers.lua:328: Found runtime errors in logfile(s): /home/runner/work/neovim/neovim/build/log/ubsan.29097
```

I tried to free the LuaRef in `LanguageTree:destroy()`, but it didn't work:
```diff
diff --git a/runtime/lua/vim/treesitter/languagetree.lua b/runtime/lua/vim/treesitter/languagetree.lua
index f274edf96..9ad89a7c5 100644
--- a/runtime/lua/vim/treesitter/languagetree.lua
+++ b/runtime/lua/vim/treesitter/languagetree.lua
@@ -527,6 +527,8 @@ function LanguageTree:destroy()
   for _, child in pairs(self._children) do
     child:destroy()
   end
+
+  self._parser:_free_logger()
 end
 
 ---@param region Range6[]
diff --git a/src/nvim/lua/treesitter.c b/src/nvim/lua/treesitter.c
index 66a75f8d4..bc08f7b76 100644
--- a/src/nvim/lua/treesitter.c
+++ b/src/nvim/lua/treesitter.c
@@ -65,6 +65,7 @@ static struct luaL_Reg parser_meta[] = {
   { "set_timeout", parser_set_timeout },
   { "timeout", parser_get_timeout },
   { "_set_logger", parser_set_logger },
+  { "_free_logger", parser_free_logger },
   { "_logger", parser_get_logger },
   { NULL, NULL }
 };
@@ -706,6 +707,25 @@ static void logger_cb(void *payload, TSLogType logtype, const char *s)
   }
 }
 
+static int parser_free_logger(lua_State *L)
+{
+  TSParser **p = parser_check(L, 1);
+  if (!p) {
+    return 0;
+  }
+
+  TSLogger logger = ts_parser_logger(*p);
+  if (!logger.log) {
+    return 0;
+  }
+
+  TSLuaLoggerOpts *opts = (TSLuaLoggerOpts *)logger.payload;
+  nlua_unref_global(opts->lstate, opts->cb);
+
+  ts_parser_set_logger(*p, (TSLogger){ .payload = NULL, .log = NULL });
+  return 0;
+}
+
 static int parser_set_logger(lua_State *L)
 {
   TSParser **p = parser_check(L, 1);
```

Any idea?

**Comment by lewis6991** [comment]
The logger doesn't need to be active for normal use.

I'll look into it.
